src/handlers/sign_tx.rs

```diff
@@ -77,3 +77,85 @@ pub fn handler(mut stream: SingleTxStream<'_>) -> Result<Signature, AppSW> {
 
     finalize_sign::end(stream, &path)
 }
+
+use ledger_device_sdk::libcall::swap::CreateTxParams;
+
+pub fn handler_swap(
+    mut stream: SingleTxStream<'_>,
+    tx_params: &CreateTxParams,
+) -> Result<Signature, AppSW> {
+    ledger_device_sdk::testing::debug_print("sign_tx.rs: handler_swap()\n");
+
+    let path = <crypto::PathBip32 as BorshDeserialize>::deserialize_reader(&mut stream)
+        .map_err(|_| AppSW::Bip32PathParsingFail)?;
+
+    ledger_device_sdk::testing::debug_print("sign_tx.rs: path computed\n");
+
+    let mut stream = HashingStream::new(stream)?;
+
+    ledger_device_sdk::testing::debug_print("sign_tx.rs: hashing stream instantiated\n");
+
+    let mut tx_prefix = parsing::types::transaction::prefix::Prefix::new();
+
+    tx_prefix
+        .deserialize_reader_in_place(&mut stream)
+        .map_err(|_err| AppSW::TxParsingFail)?;
+
+    let tx_public_key = match PublicKeyBe::try_from(tx_prefix.public_key) {
+        Ok(tx_public_key) => tx_public_key,
+        Err(_) => return Err(AppSW::PublicKeyMismatch),
+    };
+
+    let matching_private_key = {
+        let pk = ledger_device_sdk::ecc::Ed25519::derive_from_path_slip10(&path.0)
+            .public_key()
+            .map_err(|_| AppSW::KeyDeriveFail)?;
+        PublicKeyBe::from_little_endian(pk)
+    };
+
+    if tx_public_key != matching_private_key {
+        return Err(AppSW::PublicKeyMismatch);
+    }
+
+    match tx_prefix.number_of_actions {
+        1 => {
+            let action = crate::parsing::types::Action::deserialize_reader(&mut stream)
+                .map_err(|_err| AppSW::TxParsingFail)?;
+
+            match action {
+                crate::parsing::types::Action::Transfer => {
+                    let transfer = crate::parsing::types::Transfer::deserialize_reader(&mut stream)
+                        .map_err(|_err| AppSW::TxParsingFail)?;
+
+                    let amount_match = near_token::NearToken::from_yoctonear(u128::from_be_bytes(
+                        tx_params.amount,
+                    )) == transfer.deposit;
+
+                    if !amount_match {
+                        ledger_device_sdk::testing::debug_print(
+                            "sign_tx.rs: amounts do not not match\n",
+                        );
+                        return Err(AppSW::TxSignFail);
+                    }
+
```

Comment:
no check of fees ?

Reply:
They are supposed to be constant in `TGas` for a Tranfer operation (`0.450 TGas`), see https://docs.near.org/concepts/protocol/gas#cost-for-common-actions